File: 02_Two_Pointers/01_Valid_Palindrome/0125-valid-palindrome.cpp
/*
Problem: LeetCode 125 - Valid Palindrome

Description:
Given a string s, determine if it is a palindrome, considering only alphanumeric characters and ignoring cases.

Intuition:
To check if a string is a palindrome, we need to compare characters from both ends of the string. We can ignore non-alphanumeric characters and treat uppercase and lowercase letters as the same.

Approach:
1. Initialize two pointers, `left` pointing to the start of the string, and `right` pointing to the end of the string.
2. Iterate while `left` is less than `right`:
   - Skip non-alphanumeric characters by incrementing `left` or decrementing `right` if the character at that position is not alphanumeric.
   - Convert both characters to lowercase and compare them. If they are not equal, return false.
   - Increment `left` and decrement `right` to move to the next pair of characters.
3. If the loop completes without finding any non-matching characters, return true.

Time Complexity:
The time complexity is O(n), where n is the length of the input string. We iterate through the string once to check if it is a valid palindrome.

Space Complexity:
The space complexity is O(1) since we are using a constant amount of space to store the pointers and perform the comparison.
*/

class Solution {
  public:
    bool isPalindrome(string s) {
        int left = 0, right = s.length() - 1;

        while (left < right) {
            // isalnum -> Is alphabet or number
            if (!isalnum(s[left])) {
                left++;
                continue;
            }

            if (!isalnum(s[right])) {
                right--;
                continue;
            }

            if (tolower(s[left]) != tolower(s[right])) {
                return false;
            }

            left++;
            right--;
        }

        return true;
    }
};

File: 02_Two_Pointers/01_Valid_Palindrome/0125-valid-palindrome.py
"""
Problem: LeetCode 125 - Valid Palindrome

Key Idea:
To determine if a given string is a valid palindrome, we can use two pointers approach. We initialize two pointers, one at the beginning of the string (left) and the other at the end of the string (right). We then compare characters at these two pointers. If they are both alphanumeric characters and equal in value (ignoring case), we move both pointers towards the center of the string. If they are not equal, we know the string is not a palindrome. We continue this process until the two pointers meet or cross each other.

Time Complexity:
The time complexity of this solution is O(n), where n is the length of the input string 's'. In the worst case, we might need to traverse the entire string once to determine if it is a valid palindrome.

Space Complexity:
The space complexity is O(1) since we are not using any additional data structures that depend on the input size. We only use a constant amount of extra space for the two pointers and other variables.
"""


class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1

        while left < right:
            while left < right and not s[left].isalnum():
                left += 1
            while left < right and not s[right].isalnum():
                right -= 1

            if s[left].lower() != s[right].lower():
                return False

            left += 1
            right -= 1

        return True
